ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+            //make sure smart_refctd_ptr assignment ptr wont try to drop() -- .value is union
+            dst.value.constant = IR::INode::color_t(0.f);
+            std::tie(dst.value.texture.image, dst.value.texture.sampler, dst.value.texture.scale) = getTexture(src.texture);
+            if (!dst.value.texture.image) { //TODO should be using loader override here
+                assert(!dst.value.texture.sampler);
+                dst.source = IR::INode::EPS_CONSTANT;
+                dst.value.constant = IR::INode::color_t(1.f, 0.f, 0.f);//red in case when didnt find texture
+            }
+        }
+        else
+        {
+            dst.source = IR::INode::EPS_CONSTANT;
+            dst.value.constant = src.value.vvalue;
+        }
+    };
+
+    constexpr IR::CMicrofacetSpecularBSDFNode::E_NDF ndfMap[4]{
+        IR::CMicrofacetSpecularBSDFNode::ENDF_BECKMANN,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_GGX,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_PHONG,
+        IR::CMicrofacetSpecularBSDFNode::ENDF_ASHIKHMIN_SHIRLEY
+    };
+
+    IR::INode* root = ir->allocRootNode<IR::CMaterialNode>();
+    root->children.count = 1u;
+
+    bool twosided = false;
+    IR::INode::SParameter<IR::INode::color_t> opacity;
+    {
+        opacity.source = IR::INode::EPS_CONSTANT;
+        opacity.value.constant = IR::INode::color_t(1.f);
+    }
+    bool thin = false;
+
+    const CElementBSDF* current = _bsdf;
+
+    core::queue<const CElementBSDF*> bsdfQ;
+    bsdfQ.push(_bsdf);
+    core::queue<IR::INode*> nodeQ;
+    //nodeQ.push(root);
+    uint32_t childrenCountdown = 1u;
+    IR::INode* parent = root;//nodeQ.front();
+
+    while (!bsdfQ.empty())
+    {
+        current = bsdfQ.front();
+        bsdfQ.pop();
+
+        IR::INode* nextSym;
+        const auto currType = current->type;
+        switch (currType)
+        {
+        case CElementBSDF::TWO_SIDED:
+            twosided = true;
+            bsdfQ.push(current->twosided.bsdf[0]);
+            continue;
+            break;
+        case CElementBSDF::MASK:
+            getSpectrumOrTexture(current->mask.opacity, opacity);
+            bsdfQ.push(current->mask.bsdf[0]);
+            continue;
+            break;
+        case CElementBSDF::DIFFUSE:
+        case CElementBSDF::ROUGHDIFFUSE:
+            nextSym = ir->allocNode<IR::CMicrofacetDiffuseBSDFNode>();
+            getSpectrumOrTexture(current->diffuse.reflectance, static_cast<IR::CMicrofacetDiffuseBSDFNode*>(nextSym)->reflectance);
+            if (currType==CElementBSDF::ROUGHDIFFUSE)
+            {
+                getFloatOrTexture(current->diffuse.alpha, static_cast<IR::CMicrofacetDiffuseBSDFNode*>(nextSym)->alpha_u);
+                getFloatOrTexture(current->diffuse.alpha, static_cast<IR::CMicrofacetDiffuseBSDFNode*>(nextSym)->alpha_v);
+            }
+            else
+            {
+                static_cast<IR::CMicrofacetDiffuseBSDFNode*>(nextSym)->setSmooth();
+            }
```

Comment:
I noticed diffuse's `eta` doesn't get set

Reply:
because mitsuba doesnt define ior for diffuse brdf node. So it's defaulted to eta=(1,1,1) and k=(0,0,0)

Reply:
yeah I got that, so there's a default constructor then.